### PR TITLE
feat: Add peer screen

### PR DESCRIPTION
**Pull request by sdsantos**
- Android does not let me pick the .der file with any mimetype, so it let's the user pick any file.
- There's only a generic error message. Are there specific cases like an invalid certificate that we want to point out to the user?
- It's not validating the public address against the certificate (is that a possibility and, if so, should the SDK or the app validate it?)
- It does not currently support opening the app by clicking a certificate file

### CODE REVIEW

**Review by gnarea — CHANGES_REQUESTED** [review]
> Are there specific cases like an invalid certificate that we want to point out to the user?

I think it'd be good to distinguish malformed vs invalid, especially because we couldn't limit the file picker to DER files.

> It's not validating the public address against the certificate (is that a possibility and, if so, should the SDK or the app validate it?)

The SDK should do that in the future.

> It does not currently support opening the app by clicking a certificate file

Since we can't limit this to DER files, I think we can defer that requirement. The alternative would be to allow users to open any file with the Ping app, which would be annoying for them to get that option whenever they want to open a file.

---

I've found a bug: I can't create the endpoint below. [Here's its certificate](https://pong-codelab.nw.r.appspot.com/identity.der).

Screenshot_20210401-095332

I ran into this issue after creating and then deleting the `ping.awala.services` endpoint.

Logcat doesn't have anything relevant.

UPDATE: It just doesn't like the `-`, which means there are two issues here:

- Domain names containing hyphens should be allowed.
- The user should be alerted to any validation issues. And the save button should be visibly disabled.

File: app/src/main/java/tech/relaycorp/ping/ui/peers/AddPublicPeerViewModel.kt [hunk 1]
```diff
@@ -56,50 +55,69 @@ class AddPublicPeerViewModel
...
+            address.isNotBlank() && address.matches(PUBLIC_ADDRESS_REGEX)
+
+        fun isCertificatePresent() = certificateUri.isPresent
+
+        fun isValid() = isAddressValid() && isCertificatePresent()
+    }
+
+    companion object {
+        internal val PUBLIC_ADDRESS_REGEX =
+            Regex("([a-zA-Z0-9][a-zA-Z0-9-]{1,61}[a-zA-Z0-9]\\.)+[a-zA-Z]{2,}")
```
Comment (gnarea) [inline app/src/main/java/tech/relaycorp/ping/ui/peers/AddPublicPeerViewModel.kt:121]: Did you consider https://developer.android.com/reference/android/util/Patterns#DOMAIN_NAME? I think that'd be more reliable, and it's what we used in the gateway migration screen in the private gateway.

**Review by gnarea — CHANGES_REQUESTED** [review]

File: app/src/main/java/tech/relaycorp/ping/ui/peers/AddPublicPeerViewModel.kt (hunk 1 above)
Comment (sdsantos) [reply app/src/main/java/tech/relaycorp/ping/ui/peers/AddPublicPeerViewModel.kt:121]: Switched. But do note it also includes IP addresses.

Comment (gnarea) [reply app/src/main/java/tech/relaycorp/ping/ui/peers/AddPublicPeerViewModel.kt]: Sorry, I had no idea that was the case. Can you please revert it?

Comment (sdsantos) [reply app/src/main/java/tech/relaycorp/ping/ui/peers/AddPublicPeerViewModel.kt:121]: Reverted

**Comment by gnarea** [comment]
CI is failing. Will approve when it passes.

**Review by gnarea — APPROVED** [review]

**Pull request merged**